qcodes/dataset/data_set.py

```diff
@@ -961,22 +999,39 @@ def get_data_as_pandas_dataframe(self,
                 ParamSpec objects. If no parameters are supplied data for
                 all parameters that are not a dependency of another
                 parameter will be returned.
+            concat: if True individual DataFrames are concatenated along columns
+                and a single DataFrame is returned for the entire DataSet
```

Comment:
Having (public) methods that return different types is imho a code smell and should almost always be avoided. 

I think its better to have 2 different methods for this. If we really want to go this way we should use overload so that 
it is clear from the type system which type we get back

Reply:
I would suggest leaving this method as is and probably deprecate and then create 4 new methods.
for pandas.

`to_pandas_dataframe`
`to_pandas_dataframe_dict`

for xarray
`to_xarray_dataset
`to_xarray_dataarray_dict`
or something similar

Reply:
I think that suggestion to deprecate and add new methods is a good one. It always bothered me that `get_data_as_pandas_dataframe` didn't actually return a dataframe. Having a function where the `concat=True` behavior is default is great (but maybe the warning about setpoints needs to stay?). Making a distinction between `xarray.Dataset` and `xarray.DataArray` in the two suggested xarray functions is also nice.

Reply:
good point Jens and Nik, having separate methods will be much cleaner.